shared/runtime/pyexec.c

```diff
@@ -143,15 +155,41 @@ static int parse_compile_execute(const void *source, mp_parse_input_kind_t input
             mp_hal_stdout_tx_strn("\x04", 1);
         }
 
-        // check for SystemExit
-        if (mp_obj_is_subclass_fast(MP_OBJ_FROM_PTR(((mp_obj_base_t *)nlr.ret_val)->type), MP_OBJ_FROM_PTR(&mp_type_SystemExit))) {
-            // at the moment, the value of SystemExit is unused
-            ret = pyexec_system_exit;
-        } else {
+        #if MICROPY_ENABLE_VM_ABORT
+        if (nlr.ret_val == NULL) { // abort
+            ret = pyexec_abort;
+        } else
+        #endif
+        if (mp_obj_is_subclass_fast(MP_OBJ_FROM_PTR(((mp_obj_base_t *)nlr.ret_val)->type), MP_OBJ_FROM_PTR(&mp_type_SystemExit))) { // system exit
```

Comment:
This updated `SystemExit` handling code (the second commit in the PR) adds quite a bit of code size to existing builds, which don't really use the functionality.

How useful is this new code?  Does it make sense to guard it with a compile-time configuration flag?

Reply:
Or perhaps to make it even more flexible, use a macro similar to other port-specific hooks.

I.e. `MICROPY_VM_HOOK_ABORT`. This way, the code can be customized without modifying `pyexec.c`.

Reply:
I am not sure how to do the part with the hook. What do you mean, can you please explain?

Reply:
It doesnt feel like a lot of code size to me, but I can add some compile time guards, that is not a problem

Reply:
I implemented the guard using a new macro "MICROPY_ENABLE_EXIT_CODE_HANDLING"

if set to 0, the behavior should be the same as before. As well as code size. This means, the exit code is written into the variable `pyexec_system_exit` internally, before the SystemExit exception is raised (e.g. soft reset method).

if set to 1, the values defined in macros such as `PYEXEC_KEYBOARD_INTERRUPT` are used for the appropriate exit value.